orocos_kdl/tests/solvertest.cpp

```diff
@@ -1550,3 +1558,167 @@ void SolverTest::FdAndVereshchaginSolversConsistencyTest()
 
     return;
 }
+
+void SolverTest::ExternalWrenchEstimatorTest()
+{
+    /**
+     * Closed-loop test for the external wrench estimator class:
+     * Simple controlled behaviour of the robot subjected to an external force is simulated.
+     * The external wrench estimator is called in each iteration of the control loop 
+     * so to converge to final wrench value.
+     * In the end, estimated wrench is compared to the ground-truth values of the simulated wrench. 
+     */
+    std::cout << "KDL External Wrench Estimator Test" << std::endl;
+
+    int ret;
+    double eps = 0.85;
+    unsigned int nj = kukaLWR.getNrOfJoints();
+    unsigned int ns = kukaLWR.getNrOfSegments();
+    CPPUNIT_ASSERT(Equal(nj, ns)); // Current implementation of the Vereshchagin solver can only work with chains that have equal number of joints and segments
+
+    // Joint position, velocity, acceleration and torques
+    KDL::JntArray q(nj); // Joint state
+    KDL::JntArray qd(nj); // Joint state
+    KDL::JntArray q_desired(nj); // Joint state
+    KDL::JntArray qd_desired(nj); // Joint state
+    KDL::JntArray qdd(nj); // Resultant joint accelerations
+    KDL::JntArray qdd_command(nj); // Joint acceleration commands for RNE
+    KDL::JntArray command_torque(nj); // Control torque to actuate the robot
+    KDL::JntArray constraint_tau(nj); // It will result in zero in Vereshchagin for this test
+    KDL::JntArray ext_torque_reference(nj); // Ground-truth joint torques due to the external force applied on the end-effector
+    KDL::JntArray ext_torque_estimated(nj); // Estimated joint torques
+
+    // Random initial state
+    q(0) = 1.0;
+    q(1) = 0.0;
+    q(2) = 0.0;
+    q(3) = -1.57;
+    q(4) = 0.0;
+    q(5) = 1.57;
+    q(6) = -0.8;
+    q_desired = q; // desired state is the same as the initial
+
+    qd(0) = 0.2;
+    qd(1) = -0.1;
+    qd(2) = 0.3;
+    qd(3) = 0.5;
+    qd(4) = -0.1;
+    qd(5) = -0.15;
+    qd(6) = 0.9;
```

Comment:
Good question. The joint configuration that I selected here was not so randomly chosen (I need to remove the comment above), i.e. I have selected it such that we have a configuration of the robot with which we can avoid singularities for certain DoFs and be able to detect all forces that we specified. 

The necessity for that comes from the fact that the forces/moments that are applied in/around DoF that is affected by the singularity, will not be felt in the robot's joints but absorbed by the robot's structure (i.e. joint constraints) and thus, it will not be possible to estimate them (basically, it will be as that force/moment didn't exist at all). And this is not a problem with the estimator per se. but, it's just that nature does not allow for such case to be estimated... there is nothing we can do for that particular DoF. This is explained in the reference [1] that I mentioned above and also outlined in the comments in the estimator's class.

Reply:
In general, one thing that could be done is creating a set of different joint configs. for which we know that they are far-away from a singularity (w.r.t. defined external wrenches, of course). However, I am not sure about the gain from this test-extension, since the estimator is deterministic and the estimation gain is not influenced much by the joint config. (see eq. 46 in [1]).

Reply:
A set of different joint configs sounds like a good solution.

Reply:
I have added a set of 3 different (fixed-value) joint configurations. However, values for joint velocities are now randomly generated (between 0.5 and -0.5).